mbedtls/src/ssl/async_io.rs

```diff
@@ -0,0 +1,266 @@
+        }
+    }
+}
+
+pub type AsyncContext<T> = Context<AsyncIoAdapter<T>>;
+
+/// Marker type for an IO implementation that doesn't implement `tokio::io:: AsyncRead`
+/// and `tokio::io:: AsyncWrite`.
+pub enum AnyAsyncIo {}
+#[cfg(feature = "std")]
+/// Marker type for an IO implementation that implements both `tokio::io:: AsyncRead`
+/// and `tokio::io:: AsyncWrite`.
+pub enum AsyncStream {}
+
+#[async_trait]
+pub trait AsyncIo {
+    async fn recv(&mut self, buf: &mut [u8]) -> Result<usize>;
+    async fn send(&mut self, buf: &[u8]) -> Result<usize>;
+}
+
+impl<IO: AsyncIo> IoCallback<AnyAsyncIo> for AsyncIoAdapter<IO> {
+    fn recv(&mut self, buf: &mut [u8]) -> Result<usize> {
+        if let Some(cx) = unsafe { self.ecx.get() } {
+            let mut pinned_future = Box::pin(AsyncIo::recv(&mut self.inner, buf));
+            match pinned_future.as_mut().poll(cx) {
+                Poll::Ready(Ok(n)) => Ok(n),
+                Poll::Ready(Err(_)) => Err(Error::NetRecvFailed),
+                Poll::Pending => Err(Error::SslWantRead),
+            }
+        } else {
+            Err(Error::NetRecvFailed)
+        }
+    }
+
+    fn send(&mut self, buf: &[u8]) -> Result<usize> {
+        if let Some(cx) = unsafe { self.ecx.get() } {
+            let mut pinned_future = Box::pin(AsyncIo::send(&mut self.inner, buf));
+            match pinned_future.as_mut().poll(cx) {
+                Poll::Ready(Ok(n)) => Ok(n),
+                Poll::Ready(Err(_)) => Err(Error::NetSendFailed),
+                Poll::Pending => Err(Error::SslWantWrite),
+            }
+        } else {
+            Err(Error::NetSendFailed)
+        }
+    }
+}
+
+impl<IO: AsyncRead + AsyncWrite + std::marker::Unpin + 'static> IoCallback<AsyncStream> for AsyncIoAdapter<IO> {
+    fn recv(&mut self, buf: &mut [u8]) -> Result<usize> {
+        if let Some(cx) = unsafe { self.ecx.get() } {
+            let mut buf = ReadBuf::new(buf);
+            let io = Pin::new(&mut self.inner);
+            match io.poll_read(cx, &mut buf) {
+                Poll::Ready(Ok(_)) => Ok(buf.filled().len()),
+                Poll::Ready(Err(_)) => Err(Error::NetRecvFailed),
+                Poll::Pending => Err(Error::SslWantRead),
+            }
+        } else {
+            Err(Error::NetRecvFailed)
+        }
+    }
+
+    fn send(&mut self, buf: &[u8]) -> Result<usize> {
+        if let Some(cx) = unsafe { self.ecx.get() } {
+            let io = Pin::new(&mut self.inner);
+            match io.poll_write(cx, buf) {
+                Poll::Ready(Ok(n)) => Ok(n),
+                Poll::Ready(Err(_)) => Err(Error::NetSendFailed),
+                Poll::Pending => Err(Error::SslWantWrite),
+            }
+        } else {
+            Err(Error::NetSendFailed)
+        }
+    }
+}
+
+/// A `tokio::net::UdpSocket` on which `connect` was successfully called.
+///
+/// Construct this type using `ConnectedAsyncUdpSocket::connect`.
+pub struct ConnectedAsyncUdpSocket {
```

Comment:
This seems to be untested. Either remove it or implement tests.

Reply:
removed
will put all UDP logic into another PR. but I still not completely figure how to do that